core/types/meta.go

```diff
@@ -0,0 +1,100 @@
+/**
+ * Optimism 2020 Copyright
+ */
+
+package types
+
+import (
+	"bytes"
+	"encoding/binary"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+)
+
+type TransactionMeta struct {
+	L1RollupTxId      *hexutil.Uint64    `json:"l1RollupTxId"`
+	L1MessageSender   *common.Address    `json:"l1MessageSender"`
+	SignatureHashType *SignatureHashType `json:"signatureHashType"`
+}
+
+func NewTransactionMeta(L1RollupTxId *hexutil.Uint64, L1MessageSender *common.Address, sighashType *SignatureHashType) *TransactionMeta {
+	return &TransactionMeta{L1RollupTxId: L1RollupTxId, L1MessageSender: L1MessageSender, SignatureHashType: sighashType}
+}
+
+// TxMetaDecode deserializes bytes as a TransactionMeta struct.
+// The schema is:
+// varbytes(SignatureHashType) || varbytes(L1RollupTxId) || varbytes(L1MessageSender)
+func TxMetaDecode(input []byte) (*TransactionMeta, error) {
+	var err error
+	var meta TransactionMeta
+	b := bytes.NewReader(input)
+
+	sb, err := common.ReadVarBytes(b, 0, 1024, "SignatureHashType")
+	if err != nil {
+		return &TransactionMeta{}, err
+	}
+
+	var sighashType SignatureHashType
+	binary.Read(bytes.NewReader(sb), binary.LittleEndian, &sighashType)
+	meta.SignatureHashType = &sighashType
+
+	lb, err := common.ReadVarBytes(b, 0, 1024, "L1RollupTxId")
+	if err != nil {
+		return &TransactionMeta{}, err
+	}
+
+	if !bytes.Equal(lb, []byte{0x00}) {
+		var l1RollupTxId hexutil.Uint64
+		binary.Read(bytes.NewReader(lb), binary.LittleEndian, &l1RollupTxId)
+		meta.L1RollupTxId = &l1RollupTxId
+	}
+
+	mb, err := common.ReadVarBytes(b, 0, 1024, "L1MessageSender")
+	if err != nil {
+		return &TransactionMeta{}, err
+	}
+
+	if !bytes.Equal(mb, []byte{0x00}) {
+		var l1MessageSender common.Address
+		binary.Read(bytes.NewReader(mb), binary.LittleEndian, &l1MessageSender)
+		meta.L1MessageSender = &l1MessageSender
```

Comment:
Is there a reason that this and above cannot be a direct assignment?

Something like:
```
binary.Read(bytes.NewReader(mb), binary.LittleEndian, meta.L1MessageSender)
```

Reply:
I get a panic here as `meta.L1MessageSender` is `nil` since it is defined as a `*common.Address`. If I created the metadata struct with all zero values, the zero values would get written to the database.